### ISSUE TITLE
Exception when trying to start Kafka with SASL_SSL

### ISSUE THREAD

**Issue by chandanchowdhury** [bitnami/bitnami-docker-kafka#94]
**Description**
Exception when trying to start Kafka with SASL_SSL.

**Steps to reproduce the issue:**

Use a self-signed JKS with below docker-compose content.

```version: '3.5'

services:
  zookeeper:
    image: 'bitnami/zookeeper'
    container_name: zookeeper
    hostname: zookeeper
    ports:
      - 2181:2181
    environment:
      - ALLOW_ANONYMOUS_LOGIN=yes

  kafka:
    image: 'bitnami/kafka'
    container_name: kafka
    depends_on:
      - zookeeper
    ports:
      - '9093:9092'
    volumes:
      - "tmp-drive:/kafka"
    environment:
      - KAFKA_CFG_ZOOKEEPER_CONNECT=zookeeper:2181
      - KAFKA_CFG_LISTENERS=SASL_SSL://:9092
      - KAFKA_CFG_ADVERTISED_LISTENERS=SASL_SSL://:9092
      - KAFKA_BROKER_USER=user
      - KAFKA_BROKER_PASSWORD=password
      - KAFKA_CERTIFICATE_PASSWORD=password
    volumes:
      - './ssl/kafka.server.keystore.jks:/bitnami/kafka/config/certs/kafka.keystore.jks:ro'
      - './ssl/kafka.server.truststore.jks:/bitnami/kafka/config/certs/kafka.truststore.jks:ro'
```

**Describe the results you received:**

```
kafka        |  04:06:12.06 INFO  ==> Initializing Kafka...
kafka        |  04:06:12.07 INFO  ==> No injected configuration files found, creating default config files
kafka        |  04:06:12.15 INFO  ==> SASL_SSL listener detected, enabling SASL_SSL settings
kafka        | find: '/docker-entrypoint-initdb.d/': No such file or directory
kafka        |  04:06:12.32 INFO  ==> ** Kafka setup finished! **
kafka        |
kafka        |  04:06:12.34 INFO  ==> ** Starting Kafka **
kafka        | [2020-04-03 04:06:13,509] INFO Registered kafka:type=kafka.Log4jController MBean (kafka.utils.Log4jControllerRegistration$)
kafka        | [2020-04-03 04:06:13,765] INFO Registered signal handlers for TERM, INT, HUP (org.apache.kafka.common.utils.LoggingSignalHandler)
kafka        | [2020-04-03 04:06:13,765] INFO starting (kafka.server.KafkaServer)
kafka        | [2020-04-03 04:06:13,766] INFO Connecting to zookeeper on zookeeper:2181 (kafka.server.KafkaServer)
kafka        | [2020-04-03 04:06:13,776] ERROR Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer)
kafka        | org.apache.kafka.common.KafkaException: Exception while loading Zookeeper JAAS login context [java.security.auth.login.config=/opt/bitnami/kafka/conf/kafka_jaas.conf, zookeeper.sasl.client=default:true, zookeeper.sasl.clientconfig=default:Client]
version: '3.5'
kafka        | 	at org.apache.kafka.common.security.JaasUtils.isZkSecurityEnabled(JaasUtils.java:64)
kafka        | 	at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:384)
kafka        | 	at kafka.server.KafkaServer.startup(KafkaServer.scala:207)
kafka        | 	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:44)
kafka        | 	at kafka.Kafka$.main(Kafka.scala:84)
kafka        | 	at kafka.Kafka.main(Kafka.scala)
kafka        | Caused by: java.lang.SecurityException: java.io.IOException: /opt/bitnami/kafka/conf/kafka_jaas.conf (No such file or directory)
kafka        | 	at java.base/sun.security.provider.ConfigFile$Spi.<init>(ConfigFile.java:137)
kafka        | 	at java.base/sun.security.provider.ConfigFile.<init>(ConfigFile.java:102)
kafka        | 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
kafka        | 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
kafka        | 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
kafka        | 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
kafka        | 	at java.base/java.lang.Class.newInstance(Class.java:584)
kafka        | 	at java.base/javax.security.auth.login.Configuration$2.run(Configuration.java:255)
kafka        | 	at java.base/javax.security.auth.login.Configuration$2.run(Configuration.java:246)
kafka        | 	at java.base/java.security.AccessController.doPrivileged(Native Method)
kafka        | 	at java.base/javax.security.auth.login.Configuration.getConfiguration(Configuration.java:245)
kafka        | 	at org.apache.kafka.common.security.JaasUtils.isZkSecurityEnabled(JaasUtils.java:60)
kafka        | 	... 5 more
kafka        | Caused by: java.io.IOException: /opt/bitnami/kafka/conf/kafka_jaas.conf (No such file or directory)
kafka        | 	at java.base/sun.security.provider.ConfigFile$Spi.ioException(ConfigFile.java:665)
kafka        | 	at java.base/sun.security.provider.ConfigFile$Spi.init(ConfigFile.java:262)
kafka        | 	at java.base/sun.security.provider.ConfigFile$Spi.<init>(ConfigFile.java:135)
kafka        | 	... 16 more
kafka        | [2020-04-03 04:06:13,779] INFO shutting down (kafka.server.KafkaServer)
kafka        | [2020-04-03 04:06:13,783] INFO shut down completed (kafka.server.KafkaServer)
kafka        | [2020-04-03 04:06:13,783] ERROR Exiting Kafka. (kafka.server.KafkaServerStartable)
kafka        | [2020-04-03 04:06:13,785] INFO shutting down (kafka.server.KafkaServer)
kafka exited with code 1
```

**Describe the results you expected:**

Kafka should have started.

**Version**

- Output of `docker version`:

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b7f0
 Built:             Wed Mar 11 01:25:46 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 01:24:19 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

- Output of `docker info`:

```
Client:
 Debug Mode: false

Server:
 Containers: 2
  Running: 0
  Paused: 0
  Stopped: 2
 Images: 28
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.15.0-91-generic
 Operating System: Ubuntu 18.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 31.41GiB
 Name: unspecialized
 ID: G4FN:4MUM:RNQI:S5HL:2ME7:PTDE:WHOM:F2TU:QTLK:ZG3N:ICGN:IMQ4
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.25.4, build 8d51620a
docker-py version: 4.1.0
CPython version: 3.7.5
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
- Output of `lsb_release -a`
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```

**Comment by Mauraza** [comment]
Hi @chandanchowdhury ,
change 
```
 - './ssl/kafka.server.keystore.jks:/bitnami/kafka/config/certs/kafka.keystore.jks:ro'
  - './ssl/kafka.server.truststore.jks:/bitnami/kafka/config/certs/kafka.truststore.jks:ro'
```
for 
```
 - './ssl/kafka.server.keystore.jks:/opt/bitnami/kafka/config/certs/kafka.keystore.jks:ro'
 - './ssl/kafka.server.truststore.jks:/opt/bitnami/kafka/config/certs/kafka.truststore.jks:ro'
```
Tell me if this solves the problem

**Comment by chandanchowdhury** [comment]
Hey @Mauraza,
thank you for the quick response.

Tried that and got this error
```kafka        |  17:17:17.82 ERROR ==> In order to configure the SASL_SSL listener for Kafka you must mount your kafka.keystore.jks and kafka.truststore.jks certificates to the /bitnami/kafka/config/certs directory.```

**Comment by chandanchowdhury** [comment]
I think I have found the root-causes and working on a PR to fix the issue.

First one is in `run.sh` where it is still using KAFKA_HOME instead of KAFKA_CONFIGDIR when SASL is configured
```export KAFKA_OPTS="-Djava.security.auth.login.config=$KAFKA_HOME/conf/kafka_jaas.conf"```

Second, in `libkafka.sh`
* `kafka_configure_ssl_listener ` is calling `kafka_generate_jaas_authentication_file` but it should NOT
* `kafka_configure_sasl_ssl_listener` is NOT calling `kafka_generate_jaas_authentication_file` but it should

Please let me know what you think.

**Comment by juan131** [comment]
Thanks so much for the PR @chandanchowdhury !! I just approved them and I'm waiting for our CI/CD to release a new version of Kafka including them

**Comment by chandanchowdhury** [comment]
Hey @juan131,
always happy to help and big thanks to you and everyone at Bitnami for making these images available :)

We have pulled the new changes from docker-hub and the containers are up and running!!
If I don't report anything back in two days, please assume this issue resolved and can be closed.

**Comment by smoehrle** [comment]
Hey,

I ran into issues because of the changes made by the associated pull request. I want to run kafka with a custom kafka_jaas.conf. This worked fine until recently. But now it is no longer possible.

**In detail:**
I have a kafka_jaas.conf file which I am mounting into my container. If I mount it to 

`/opt/bitnami/kafka/config/kafka_jaas.conf`

(the place where it is inside of the container), it gets overwritten on every start-up because of the changes in `2/debian-10/rootfs/opt/bitnami/scripts/libkafka.sh` made by the pull-request. If I mount it somewhere else and specify the place in the env var 

`KAFKA_OPTS=-Djava.security.auth.login.config=/config/kafka_jaas.conf`

then the container still uses the default place because KAFKA_OPTS is overwritten in the run.sh script.

**Possible solutions I can think of:**
1. Do not recreate kafka_jaas.conf when it already exists. (Possibly not desired)
2. Do not overwrite KAFKA_OPTS if it is already defined

**Comment by Mauraza** [comment]
Hi @smoehrle, 

You are mounting in the wrong path, The right path is `/bitnami/kafka/conf`.
if you want more information you can see it in [Full configuration](https://github.com/bitnami/bitnami-docker-kafka#full-configuration).
I hope that helps you.

**Comment by smoehrle** [comment]
That's not the issue. I already tried that. The only thing that is different when I mount to `/opt/bitnami/kafka/conf` is the following log entry:

```
broker_1     |  12:08:31.93 WARN  ==> Detected files mounted to /opt/bitnami/kafka/conf. This is deprecated and files should be mounted to /bitnami/kafka/config.
```

I also tried `/bitnami/kafka/config` (as in the log message) or `/bitnami/kafka/conf` (as you stated).

The problem is that run.sh hard codes the path for the `kafka_jaas.conf` to `/opt/bitnami/kafka/config/kafka_jaas.conf` and that file is regenerated on every startup.

**Comment by Mauraza** [comment]
You don't have to run again the container, just restart. [It's step 3](https://github.com/bitnami/bitnami-docker-kafka#step-3-restart-kafka), because with run you create a new container with the run.sh other-ways with the command start (container) or restart (docker-compose) launch a container previously stopped so the run.sh not will be executed.

**Comment by smoehrle** [comment]
I don't have to, but I want to. I want to use `docker-compose down` and `docker-compose up -d` without things breaking. I don't want any state in the container which I will lose, when the container is gone.

In my opinion the problem is not that run.sh runs on every start-up. The problem is that I can set the env var KAFKA_OPTS as implied in many places in the documentation but it has no effect on start-up because it is overwritten.

**Comment by chandanchowdhury** [comment]
@smoehrle The changes I made was to fix an known issue where SASL_SSL was not working as intended, not to break what you were doing.

Leaving that aside, we have multiple options to fix the side effect and/or achieve what you need. 
However, first we need to know what is making you include "_kafka_jaas.conf_" using KAFKA_OPTS instead of using environment variables to generate the config which was the original intent of _bitnami/kafka_ image? An example of your _kafka_jaas.conf_, without the sensitive details of course, would be great help.

**Comment by smoehrle** [comment]
Your changes seem completely reasonable to me. Seems like we relied on that behavior/bug. But to make it a complete story. I inherited the project a few days ago because we needed to tweak a few things due to some unrelated issues. After a docker image update the previous working configuration did not work any more. The SASL authentication made the most trouble. I tried the environment variables to generate the config. But since we have only one instance I only set the `KAFKA_BROKER_USER` and not the `KAFKA_INTER_BROKER_USER`. Since I was still able to log in with the default credentials (user/bitnami) I abandoned that path. I didn't know about the file generation back then and I knew that it was possible (due to the bug) to mount a custom kafka_jaas.conf. Pursuing that path I finally landed here.

**So long story short:** Previously it was possible to mount a custom kafka_jaas.conf because of the wrong path in run.sh. It would be nice if we can come up with a solution so make that still possible. Our use case is the desire for multiple users and not just one. But since it was previously possible I can imagine that more people with different use-cases rely on that behavior

**Comment by chandanchowdhury** [comment]
After update things started breaking, sound like story of our life :D

For put off the immediate fire, do you think you can use {{bitnami/kafka:2.4.1-debian-10-r43}} tag of the image?
If so, we should close this issue and open a new one to request support for multiple SASL username and password. The implementation can be complex or simple based on how it will be implemented, but we should discuss with the repo maintainers first.

**Comment by smoehrle** [comment]
My workaround is a modified run.sh which I mount into the container :)

I created a new issue (#98). This on can be closed from my side

**Comment by Mauraza** [comment]
I close the issue as a PR has been created
